src/tests/testClpMap.cc

```diff
@@ -153,55 +165,41 @@ testClpMap::testConstructor()
 }
 
 void
-testClpMap::testSetMemLimit()
+testClpMap::testMemoryLimit()
 {
-    TestMap m(2048);
-    // overflow the map with entries to make sure it has lots of entries to purge below
-    fillMapWithElements(m);
-    auto testEntriesBefore = m.entries();
-    CPPUNIT_ASSERT(testEntriesBefore > 0);
-
-    m.setMemLimit(m.memoryUsed() / 2);
-    auto entriesAfterPurge = m.entries();
-    CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
-
-    m.setMemLimit(m.memLimit() * 2);
-    // overflow the map with entries again to make sure it can grow after purging
+    const size_t initialCapacity = 1024; // bytes
+    TestMap m(initialCapacity);
     fillMapWithElements(m);
-    CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
-
-    // progressively shrink to size 0, checking that the number of
-    // entries decreases each time
-    testEntriesBefore = m.entries();
-    const auto singleEntrySizeEstimate = *TestMap::MemoryCountedFor("0", 0) + 10;
-    while (m.memoryUsed() > singleEntrySizeEstimate) {
-        m.setMemLimit(m.memoryUsed() / 2);
-        entriesAfterPurge = m.entries();
-        CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
-        CPPUNIT_ASSERT(entriesAfterPurge > 0);
-        testEntriesBefore = entriesAfterPurge;
-    }
+    const auto entriesAtInitialCapacity = m.entries();
 
-    // set size to 0, all entries should have been removed
+    // check that all entries are removed if we prohibit storage of any entries
     m.setMemLimit(0);
     CPPUNIT_ASSERT_EQUAL(size_t(0), m.entries());
 
-    // test that memory used decreases when shrinking
-    size_t curLimit = 10240;
-    size_t lastMemoryUsed;
-    m.setMemLimit(curLimit);
+    // test whether the map can grow after the all-at-once purging above
+    m.setMemLimit(initialCapacity * 2);
     fillMapWithElements(m);
-    lastMemoryUsed = m.memoryUsed();
-    while (curLimit > 1024) { // stop at any practical size
-        CPPUNIT_ASSERT(m.memoryUsed() < curLimit);
-        CPPUNIT_ASSERT(m.memoryUsed() <= lastMemoryUsed);
-        // also check that we can still add entries - evicting old ones if needed
-        CPPUNIT_ASSERT(m.add(std::to_string(curLimit+1000), 1, 10));
-        CPPUNIT_ASSERT_EQUAL(1, *m.get(std::to_string(curLimit + 1000)));
-        curLimit /= 2;
-        lastMemoryUsed = m.memoryUsed();
-        m.setMemLimit(curLimit);
+    CPPUNIT_ASSERT(m.entries() > entriesAtInitialCapacity);
+
+    // test that memory usage and entry count decrease when the map is shrinking
+    while (m.entries()) {
+        // also check that we can still add entries, evicting old ones if needed
+        // (at least as long as there is at least one entry in the map)
+        addOneEntry(m, 0);
```

Comment:
Note that adding an entry fails with error if the map storage is too small to even contain one entry

Reply:
You are right (and I came to the same conclusion after leaving the office last night). Replaced with a TODO in 42708cf.